[Blog] The HTML-first approach of Glimmer components (#1293)
simplabs/simplabs.github.io: _posts/2020-09-28-html-first-gllimmer-components.md

---

> title: 'The HTML-first approach of Glimmer components'

Reviewer: Is this a fitting title? HTML-first is really only one point in the post after all and comes more at the end. Maybe something like

```suggestion
title: 'How Glimmer components makes working with components in Ember simpler and more powerful'
```

or so?

Author: "Simpler and more powerful components in Ember Octane with Glimmer Components"?

Reviewer: sounds good to me 👍

author: 'Ricardo Mendes'
github: locks
twitter: locks
topic: ember
bio: 'Senior Frontend Engineer, Ember Framework and Learning Core teams member'
description:
  'Ricardo Mendes explains how Glimmer components simplify Ember with an
  HTML-first approach'
og:
  image: /assets/images/posts/2020-08-28-testing-the-miragejs-setup/og-image.png
---

Following up on
[Bringing clarity to templates through Ember Octane](../../../../2020/09/28/clarity-in-templates/),
we will be discussing how the Glimmer components introduced by the
[Ember Octane edition](https://blog.emberjs.com/2019/12/20/octane-is-here.html)
aim to modernize and simplify Ember applications by using native JavaScript
syntax and an HTML-first approach.

<!--break-->

![Bringing clarity to your Ember templates](/assets/images/posts/2019-12-20-clarity-in-templates/illustration.jpg#full@720-1440)

## Introducing Glimmer components

The release of the
[Ember Octane edition](https://blog.emberjs.com/2019/12/20/octane-is-here.html)
back in December was one of Ember's biggest releases, bringing with it modern

> and streamlined APIs. At the core of the release are tracked properties and

Reviewer: could we add a link to the guide for tracked properties?

Author: Added.

> Glimmer components, which go hand in hand for reasons you will see later. While

Reviewer: I think it could just be leaved out?

> Octane, and subsequently Glimmer components, has been out for quite some time,

Reviewer: ```suggestion
Octane, and subsequently Glimmer components, have been out for quite some time,
```

?

Author: Re-ordered for clarity.

> you are likely using Ember (or classic) components in your application. To give

Reviewer: ```suggestion
you are likely using Ember (or _"classic"_) components in your application. To give
```

I think that makes it clearer what you're trying to say here.

some context as to the impact of Glimmer components in the Ember mental model,

> I'll be introducing Glimmer components as they relate to classic components.

Author: Re-wrote this sentence. Let me know if it's clearer @nickschot.

The key points we will be addressing in this post are:

> - Glimmer components do not extend from EmberObject, and use the native `class`

Reviewer: ```suggestion
- Glimmer components do not extend from `EmberObject`, and use the native `class`
```

  syntax.
- Separation of internal state (properties) and external state (arguments).
- HTML-first approach makes for much simpler API surface.

## Native `class`

The first thing that jumps to the eye is that Glimmer components have a
different base class with a radically different API (see
[Ember Component](https://api.emberjs.com/ember/3.21/classes/Component) vs
[Glimmer component](https://api.emberjs.com/ember/3.21/modules/@glimmer%2Fcomponent)).

With the new Glimmer component implementation you are expected to use the native
`class` syntax, but it also changes something more important: a Glimmer

> component no longer inherits from EmberObject. Now you should use native getters

Reviewer: ```suggestion
component no longer inherits from `EmberObject`. Now you should use native getters
```

> component no longer inherits from EmberObject. Now you should use native getters

Reviewer: ```suggestion
component no longer inherits from EmberObject. Now you can use native getters
```

sounds more positive?

and setters, as well as tracked properties.

> How you define component actions also suffers an adjustment. First, a bit of

Reviewer 2: Not sure if "suffers" is the right word here? "requires" maybe?

historical context. One of the first proposals for Ember components had actions
defined as a member function of the component, like so:

```javascript
export default Component.extend({
  myActionHandler() {
    console.log('myActionHandler triggered');
  },
});
```

Unfortunately, due to the API surface of components and the fact that one of the
component lifecycle hooks was named `destroy`, that meant that users would
unknowingly override component hooks and introduce strange bugs to their
application. To address this, a system was conceived where you define component
actions in the actions hash:

```javascript
export default Component.extend({
  actions: {
    myActionHandler() {
      console.log('myActionHandler triggered');
    },
  },
});
```

> But now that Glimmer components have a quite a small API (roughly `willDestroy`,

Reviewer 2: ```suggestion
But now that Glimmer components have quite a small API surface (roughly `willDestroy`,
```

> But now that Glimmer components have a quite a small API (roughly `willDestroy`,

Reviewer: ```suggestion
But now that Glimmer components have a quite a small API (essentially the `willDestroy`,
```

> `isDestroying` and `isDestroyed`), we can go back to defining them as methods in

Reviewer: ```suggestion
`isDestroying` and `isDestroyed` methods only), we can go back to defining them as methods in
```

> the component, and then refer to them directly in the template.

Reviewer 2: So my comma knowledge in English is a bit rusty, but I think a comma is not necessary before "and"
```suggestion
the component and then refer to them directly in the template.
```

> the component, and then refer to them directly in the template.

Reviewer: ```suggestion
the component, and then refer to them directly in the template without risking unintentional overrides.
```

?

component is invoked like so:

```handlebars
<OrderPasta @pasta="Spaghetti" @sauce="Carbonara" />
```

Now let us look at how both implementations differ. First, the Ember component:

```javascript
/// app/components/order-pasta.js
import Component from '@ember/component';

export default Component.extend({
  dishName: computed('pasta', 'sauce', function () {
    return `${this.pasta} ${this.sauce}`;
  }),

  actions: {
    handleReservation() {
      console.log(`Ordered a plate of ${this.dishName}`);
    },
  },
});
```

```handlebars
<div>{{this.dishName}} <button {{action 'handleReservation'}}>Reserve</button></div>
```

And now the Glimmer component:

```javascript
// app/components/order-pasta.js
import Component from '@ember/component';
import { tracked } from '@glimmer/tracking';
import { action } from '@ember/object';

export default class OrderPasta extends Component {
  get dishName() {
    return `${this.args.pasta} ${this.args.sauce}`;
  }

  @action
  handleReservation() {
    console.log(`Ordered a plate of ${this.dishName}`);
  }
}
```

```handlebars
<div>{{this.dishName}} <button {{on 'click' this.handleReservation}}>Reserve</button></div>
```

> I won't go into default on the `action` to `on` modifier changes, but notice how

Reviewer 2: ```suggestion
I won't go into detail on the `action` to `on` modifier changes, but notice how
```

in the Glimmer component we reference the callback directly, we used a native

> getter, and we defined the action handler as a decorated method. You might also

Reviewer 2: Same thing with the ", and" here as above.
```suggestion
getter and we defined the action handler as a decorated method. You might also
```

have noticed that in the Glimmer component code I have referred to the passed in
values as `this.args`, which brings us to our next point.

## Separation of properties and arguments

One behavior of classic components that can sneak up on even more experienced
developers is the fact that classic components reflect passed-in state
(arguments) onto internal state (properties). This has a big implication. Since
properties and arguments are directly linked in classic components, you can
effectively override any property from the outside by calling the component with
an argument with the same name. You cannot refer to the argument's original

> value from within the component.

Reviewer: Would this benefit of some added information about what the negative consequences of that are?

Glimmer components do away with this behavior by separating arguments into the
`args` property of the component, and by making named argument bindings
read-only, which gives developers a better guarantee that the data flow in a

> component tree isn't accidentally two-way bound.

Reviewer: Would this benefit of some added information about what the negative consequences of two-way bindings are?

is enough. If the component is invoked with a different `pasta` argument, this
will be overwritten.

```javascript
/// app/components/order-pasta.js
import Component from '@ember/component';

export default Component.extend({
  pasta: 'Spaghetti',
});
```

For the Glimmer component we have to do a bit more legwork. We'll assume that
`pasta` can change over time, so we will add a native getter so it gets
automatically updated. We will also need to change our `dishName` getter to
refer to the property instead.

```javascript
// app/components/order-pasta.js
export default class OrderPasta extends Component {
  get pasta() {
    return this.args.pasta || 'Spaghetti';
  }

  get dishName() {
    return `${this.pasta} ${this.args.sauce}`;
  }
}
```

As you can see, we can now reason about the code locally, instead of having to
rely on the framework knowledge that `pasta` would be overwritten. We can use
the same name, `pasta`, because there's a separation of properties and
arguments, but there's no ambiguity as it's explicit in the class what is a
property (`this.something`) and what is an argument (`this.args.something`),

> like `{{this.something}}` and `{{@something}}` in the template.

Reviewer 2: I would somewhat rewrite this sentence to make the transition from "as it's explicit in the class what is a property" smoother.

Maybe: "as what is a property ... and what is an argument ... is explicitly defined in the class"

Reviewer: Would it also make things clearer to use `this.pasta`, `this.args.pasta` etc. vs. `something`?

Author: Re-wrote this in a more direct way. I think it is clearer now.

If you checked the API documentation linked at the beginning of the blog post,
you might still be wondering how Glimmer components managed to reduce the API
surface so much. We will not be covering lifecycle hooks in this post (which
account for the removal of `didInsertElement`, `didReceiveAttrs`, `didRender`,
`didUpdate`, `didUpdateAttrs`, `willRender`, `willUpdate`), instead we will be
focusing on the APIs that allows one to customize the wrapper element of Ember
Components.

## HTML-first APIs

Ember components have been around since before the 1.0 release, where they were
introduced as isolated
[views](https://api.emberjs.com/ember/1.0/classes/Ember.View). Views are long
gone from everyday applications, but their legacy still lives on in Ember

> component. Ember components have a very distinct characteristic, the component's

Reviewer 2: ```suggestion
components. Ember components have a very distinct characteristic, the component's
```

template is wrapped by a hidden element, `<div>` by default. To customize this
element, you have to use certain APIs in the component's class, like `tagName`,
`classNameBindings` and `attributeBindings`.

Let us go with a practical example to make it clearer. We are going to make a

> button component that receive a `primary` CSS class, and can receive a

Reviewer 2: ```suggestion
button component that receives a `primary` CSS class, and can receive a
```

`disabled` HTML attribute. The component template itself will be the content for
the button.

```javascript
// app/components/my-button.js
import Component from '@ember/component';

export default Component.extend({
  tagName: 'button',

  attributeBindings: ['isDisabled:disabled'],
  isDisabled: false,

  classNameBindings: ['buttonType'],
  buttonType: 'primary',
});
```

```handlebars
// app/components/my-button.hbs
{{yield}}
```

Now you can call `MyButton` and dynamically change the class and the attribute:

```handlebars

> <MyButton>Click me</MyButton>                                                   {{!-- renders --}} <button class="primary">Click me</button>

Reviewer: ```suggestion
<MyButton>Click me</MyButton>
{{!-- renders <button class="primary">Click me</button>  --}}
```

You won't see the comment when the line is super long. I'd change the others like this as well.

longer have an implicit wrapper element so now you can configure this in the
component's template itself. This, coupled with being able to specify HTML
attributes and the new `...attributes` make template-only components much more
feasible.

Let's implement `MyButton` in Glimmer. As mentioned, we are moving the code from
the component class to the template, so we don't need the JavaScript file.

```handlebars
// app/components/my-button.hbs
<button
  class={{if @buttonType @buttonType 'primary'}}
  ...attributes
>
  {{yield}}
</button>
```

Much more straightforward! In `{{if @buttonType @buttonType 'primary'}}` we are
saying that the value should be `primary` if `@buttonType` is not defined, and
in `...attribute` we are telling Ember where to put any HTML attributes that are
passed in when calling the component.

In this case we do not want to pass `class` as an HTML attribute when calling
the component because `class` is special in that it merged together the existing
classes in the component with whatever you pass to the component:

```handlebars
// app/components/multiple-classes.hbs
<div class="one" ...attributes></div>
```

```handlebars

> <MultipleClasses class="two" />

Reviewer: Would it be easier to use the above button as an example here?

```

Renders:

```html
<div class="one two"></div>
```

For the sake of exemplifying `attributeBindings` I glossed over the fact that
Ember will apply any HTML attributes you pass to an Ember component to the
implicit wrapper element. Now that we are explicitly using `...attributes` in
our Glimmer component, we need to update how we're calling the component:

```handlebars

> <MyButton>Click me</MyButton>                                               {{!-- renders --}} <button class="primary">Click me</button>

Reviewer: ```suggestion
<MyButton>Click me</MyButton>
{{!-- renders <button class="primary">Click me</button> --}}
```

see above – I'd change like this for all of the invocations.

Author: I moved the "renders" to a newline but kept the markup outside of the handlebars comment so they get syntax highlighting.

<MyButton disabled={{true}}>Click me</MyButton>                             {{!-- renders --}} <button class="primary" disabled>Click me</button>
<MyButton @buttonType={{'secondary'}}>Click me</MyButton>                   {{!-- renders --}} <button class="secondary">Click me</button>
<MyButton @buttonType={{'secondary'}} disabled={{true}}>Click me</MyButton> {{!-- renders --}} <button class="secondary" disabled>Click me</button>
```

A common point of dissatisfaction with frameworks, including Ember, is the so

> called "magic". This is when something the framework does something for you that

Reviewer: ```suggestion
called "magic". This is when something the framework does for you that
```

you have no insight into, so you may not understand that it is happening or why.
For example, Ember components having an implicit wrapper element, and Ember
automatically applying `...attributes` to it.

With the introduction of Glimmer components in the Octane edition plus the
template improvements already mentioned in a previous post, Ember has made it
much easier to reason locally about your component. By keeping layout logic in
the layout, instead of the component class you only have to look in one place to
know what the component will render. By moving towards native JavaScript syntax

> and functionality, Ember has diminished the uncertain of whether a certain

Reviewer 2: ```suggestion
and functionality, Ember has diminished the uncertainty of whether a certain
```

functionality is provided by Ember or by JavaScript.

I hope this exploration of Glimmer components and Ember components was useful

> for you, and at the very least gave you a renewed appreciation for the Octane

Reviewer 2: ```suggestion
for you and at the very least gave you a renewed appreciation for the Octane
```

---
title: ''
author: 'Ricardo Mendes'
github: locks
twitter: locks
topic: ember
bio: 'Senior Frontend Engineer, Ember Framework and Learning Core teams member'
description:
  'Ricardo Mendes explains how Glimmer components simplify Ember with an
  HTML-first approach'
og:
  image: /assets/images/posts/2020-08-28-testing-the-miragejs-setup/og-image.png
---

Following up on

> [Bringing clarity to templates through Ember Octane](../../../../2020/09/28/clarity-in-templates/),

Reviewer: The link doesn't seem to be correct. I'd recommend using an absolute link `/blog/2020/…`

Author: Link has been fixed.

focused.

For the Ember component, we can see that adding a property with the proper name
is enough. If the component is invoked with a different `pasta` argument, this
will be overwritten.

```javascript
/// app/components/order-pasta.js
import Component from '@ember/component';

export default Component.extend({
  pasta: 'Spaghetti',
});
```

For the Glimmer component we have to do a bit more legwork. We'll assume that
`pasta` can change over time, so we will add a native getter so it gets
automatically updated. We will also need to change our `dishName` getter to
refer to the property instead.

```javascript
// app/components/order-pasta.js
export default class OrderPasta extends Component {
  get pasta() {
    return this.args.pasta || 'Spaghetti';
  }

  get dishName() {
    return `${this.pasta} ${this.args.sauce}`;
  }
}
```

As you can see, we can now reason about the code locally instead of having to
rely on the framework knowledge that `pasta` would be overwritten. We can use
the name `pasta` for both the property and the argument because the syntax makes
it clear which is which. If you need access to the argument you do

> `this.args.pasta` in the class or `{{@pasta}}` in the template. If you need to

Reviewer: This is not really a comment on the post but is there any difference between `{{@pasta}}` and `{{this.args.pasta}}`?

Author: There is a slight difference. `{{@pasta}}` is directly the reference that was passed to the component, while `{{this.args.pasta}}` will only work if you have a JavaScrip instance (so no template-only components), and the component has to read the reference and then wrap it again to pass it into the instance `args`.
